### PR TITLE
Client option to get response in df_to_inventory

### PR DESCRIPTION
**Pull request by d-chambers**
This PR adds the option to use a `get_stations_kwargs` row in a dataframe passed to `obsplus.utils.stations.df_to_inventory` in order to download an existing channels client with the default ObsPy fdsn client or one provided to the function.

For context, it is nice to keep a simple csv of channels and be able to edit it and regenerate an inventory rather than work with the highly nested inventory object directly. This adds just one more optional column that should make it easier to get the correct response for channels registered with an FDSN provider.

### CODE REVIEW

**Comment by d-chambers** [comment]
@sboltz mind taking a look at this? I know `df_to_inventory` is a but weird with so many nested functions but otherwise should be straightforward.

**Comment by sboltz** [comment]
Sure, I can take a look. Thanks for implementing it.

**Comment by d-chambers** [comment]
No problem, it seemed like the right thing to do.

File: obsplus/utils/stations.py [hunk 1]
```diff
@@ -38,15 +49,42 @@ def df_to_inventory(df) -> obspy.Inventory:
     df
         A dataframe which must have the same columns as the once produced by
         :func:`obsplus.stations_to_df`.
```
Comment (sboltz) [inline obsplus/utils/stations.py:51]: I might recommend actually listing the columns here (network, station, location, channel, seed_id, latitude, longitude, elevation, depth, start_date, end_date, dip, azimuth, sampling_rate (sample_rate?), if you don't have them handy).

I had to actually go convert an inventory to a dataframe to figure out what was required here.

File: obsplus/utils/stations.py [hunk 2]
```diff
@@ -38,15 +49,42 @@ def df_to_inventory(df) -> obspy.Inventory:
...
+    client
```
Comment (sboltz) [inline obsplus/utils/stations.py:52]: This needs to be filled in...
Perhaps "a client that can be used to download station response information"?

File: obsplus/utils/stations.py [hunk 3]
```diff
@@ -38,15 +49,42 @@ def df_to_inventory(df) -> obspy.Inventory:
...
     Notes
     -----
-    The dataframe can also contain columns named "sensor_keys" and
-    "datalogger_keys" which will indicate the response information should
-    be fetched suing obspy's ability to interact with the nominal response
-    library. Each of these columns should either contain tuples or strings
-    where the keys are separated by double underscores (__).
+    There are two ways to include response information:
+
+    1. Using the Nominal Response Library (NRL):
```
Comment (sboltz) [inline obsplus/utils/stations.py:59]: I like the added detail here

File: obsplus/utils/stations.py [hunk 4]
```diff
@@ -38,15 +49,42 @@ def df_to_inventory(df) -> obspy.Inventory:
...
+    If the dataframe contains a column get_stations_kwargs it indicates that
+    either a client was passed as the client argument or the fdsn IRIS client
+    should be used to download station information. The contents of this column
+    must be a dictionary of acceptable keyword arguments for the client's
+    `get_stations` method. All time values must be provided as iso8601
+    strings.
+    For example,
+    {'network': 'UU', 'station': 'TMU', 'location': '01', 'channel': 'HHZ',
+     'starttime': '2017-01-02',}
+    would use either a provided station client (or ObsPy's default FDSN client
```
Comment (sboltz) [inline obsplus/utils/stations.py:82]: Would it be more concise to say "would be passed to the station client"? You already explain the choices for the client earlier.

File: obsplus/utils/stations.py [hunk 5]
```diff
@@ -114,19 +164,74 @@ def _get_resp_key(key):
             return tuple(key)
 
     def _maybe_add_response(series, channel_kwargs):
-        """ Maybe add the response information if required columns exist. """
+        """ Simple dispatch to determine which response function to use."""
+        # if no response columns are found exit early
+        if response_cols.isdisjoint(set(series.index)):
+            return
+        _maybe_add_nrl_response(series, channel_kwargs)
+        if "response" not in channel_kwargs:
```
Comment (sboltz) [inline obsplus/utils/stations.py:172]: Am I understanding correctly that "response" gets added if it succeeds in fetching the NRL response?

File: tests/test_utils/test_stations_utils.py [hunk 6]
```diff
@@ -190,3 +126,139 @@ def test_00_location_code(self, df_from_inv):
...
+        return df_with_nrl_response
+
+    @pytest.fixture
+    def df_with_both_response_cols(
+        self, df_with_nrl_response, df_with_get_stations_kwargs
+    ):
+        """
+        Get a df with both types of responses.
+
+        row 1 should have NRL, row 2 both, and row three client kwargs
```
Comment (sboltz) [inline tests/test_utils/test_stations_utils.py:223]: Am I understanding correctly that row 2 has both the information to fetch the nominal response and to retrieve from a provided client? If so, how does one know which set of arguments take precedence?

**Comment by d-chambers** [comment]
Thanks @sboltz, all good points. I think I addressed all of them and I also couldn't resist the urge to refactor a little more. I will give it one more look-over tomorrow.

File: tests/test_utils/test_stations_utils.py (hunk 6 above)
Comment (d-chambers) [reply tests/test_utils/test_stations_utils.py:226]: I added a note about this in the docs, I think the thing to do is just give the NRL precedence.

**Comment by d-chambers** [comment]
I should note this has one small breaking change:
For the datalogger_keys and sensor_keys columns, obsplus now expects json strings or python objects. Separating the keys with __ was kinda weird, and a mistake in my opinion.

**Comment by sboltz** [comment]
> I should note this has one small breaking change:
> For the datalogger_keys and sensor_keys columns, obsplus now expects json strings or python objects. Separating the keys with __ was kinda weird, and a mistake in my opinion.

I agree that that was a kind of weird way to do it (though convenient when you are too lazy to figure out the 'right' way to do it). I think taking that out makes sense.

File: tests/test_utils/test_stations_utils.py (hunk 6 above)
Comment (sboltz) [reply tests/test_utils/test_stations_utils.py:226]: All right. I don't have a strong opinion on this, but one thing to consider: should the station client actually take precedence over the NRL? That could theoretically have more specific information associated with it. Ideally, only one or the other would be specified for each station/channel and this doesn't really matter, though.

**Comment by d-chambers** [comment]
> In the face of ambiguity, refuse the temptation to guess.

I succumbed to the temperation to guess. It should probably raise an error if both NRL and station client are specified.

**Pull request merged**